[MediaRecorder] Allow opus in mp4 when creating a MediaRecorder

jyavenard opened the pull request:
#### 9f032832a4e2bb7f52643588071da7d3954b6b09
<pre>
[MediaRecorder] Allow opus in mp4 when creating a MediaRecorder
<a href="https://bugs.webkit.org/show_bug.cgi?id=282220">https://bugs.webkit.org/show_bug.cgi?id=282220</a>
<a href="https://rdar.apple.com/138811924">rdar://138811924</a>

Reviewed by Youenn Fablet.

Allow Opus in MP4; the MediaRecorderPrivateEncoder already supported the feature.

Added tests. We split the existing tests to reduce the time it takes to run them and also only test for two bitrates.
webkit.org/b/282322 shows that with two many tests running at once the MediaRecorder stops getting data from the MediaStream.
To be further investigated.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-mp4-opus-expected.txt: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-mp4-opus.html: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-webm-expected.txt: Copied from LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-expected.txt.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-webm.html: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate.html:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate.js: Added.
(record):
* LayoutTests/http/wpt/mediarecorder/mimeType.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
(WebCore::MediaRecorderPrivateAVFImpl::isTypeSupported):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
(WebCore::MediaRecorderPrivateWriterAVFObjC::close): FlyBy: calling markAsFinished should be call from the serial queue passed to -requestMediaDataWhenReadyOnQueue:usingBlock: which we do not use. Remove the use.

Canonical link: <a href="https://commits.webkit.org/285952@main">https://commits.webkit.org/285952@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f7d69467a32afc35f5bf3f13db8d728a3416e61

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74289 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53718 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27100 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78668 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25527 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76406 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1503 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58399 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16730 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77356 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48563 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63907 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38811 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45531 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21402 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23860 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67426 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21749 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80185 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73547 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/920 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66691 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1751 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63925 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65971 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16371 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9912 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8066 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/95328 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1570 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4358 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20933 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1599 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1587 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1606 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate.html:
```diff
@@ -9,7 +9,7 @@
     function record(stream, bitRate, mimeType)
     {
-        const recorder = new MediaRecorder(stream, { mimetype: mimeType, audioBitsPerSecond : bitRate });
+        const recorder = new MediaRecorder(stream, { mimeType, audioBitsPerSecond : bitRate });
```
I don't understand this change. This function is passed a mime type string so your new test will pass something like

    { 'audio/mp4; codecs=opus', audioBitsPerSecond: 128000 }

to the `MediaRecorder` constructor, but the `options` parameter to the `MediaRecorder` constructor is a dictionary so it should pass this

    { mimetype: 'audio/mp4; codecs=opus', audioBitsPerSecond: 128000 }

jyavenard replied:
the option to be given is `mimeType` ; there was a typo. It was written `mimetype`. If you pass in the dictionary a parameter that has the same name as the key, it's redundant.
e.g. this is equivalent to: 
`const recorder = new MediaRecorder(stream, { mimeType: mimeType, audioBitsPerSecond : bitRate });`

jyavenard replied:
I've made it more explicit with `mimeType: mimeType`

for the record.
```
var mimeType = 'video/webm'
var dictionary = { foo:1, mimeType }
```
dictionary -> `{foo: 1, mimeType: "video/webm"} = $4`

jyavenard committed:
[MediaRecorder] Allow opus in mp4 when creating a MediaRecorder
https://bugs.webkit.org/show_bug.cgi?id=282220
rdar://138811924

Reviewed by Youenn Fablet.

Allow Opus in MP4; the MediaRecorderPrivateEncoder already supported the feature.

Added tests. We split the existing tests to reduce the time it takes to run them and also only test for two bitrates.
webkit.org/b/282322 shows that with two many tests running at once the MediaRecorder stops getting data from the MediaStream.
To be further investigated.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-expected.txt:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-mp4-opus-expected.txt: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-mp4-opus.html: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-webm-expected.txt: Copied from LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-expected.txt.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate-webm.html: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate.html:
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-bitrate.js: Added.
(record):
* LayoutTests/http/wpt/mediarecorder/mimeType.html:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateAVFImpl.cpp:
(WebCore::MediaRecorderPrivateAVFImpl::isTypeSupported):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterAVFObjC.mm:
(WebCore::MediaRecorderPrivateWriterAVFObjC::close): FlyBy: calling markAsFinished should be call from the serial queue passed to -requestMediaDataWhenReadyOnQueue:usingBlock: which we do not use. Remove the use.

Canonical link: https://commits.webkit.org/285952@main
